### ISSUE TITLE
[imtcp] encrypted transmission use a plaintext key, it is not safe.

### ISSUE THREAD

**Issue by wsp1991** [rsyslog/rsyslog#4244]
### Expected behavior
pre action: openssl genrsa -out server.key 2048
when imtcp module use server.key to encrypted transmission，
the server.key is a plaintext key. it need to require encryption.

### Actual behavior
the server.key is a plaintext key, There is a security risk.
### Environment
- rsyslog version: 
8.2002.0
- platform: 
NA
- for configuration questions/issues, include rsyslog.conf and included config files
NA

**Comment by davidelang** [comment]
If you encrypt the key, where do you store the encryption key for the key? in 
plaintext in the rsyslog.conf?

how is that really any more secure?

that said, a lot of companies have 'checkbox security' that won't care about the 
actual security, they just will have a mandate to not have keys with no 
passwords, so rsyslog should support this as an option.

David Lang

  On Tue, 14 Apr 2020, 
wsp1991 wrote:

**Comment by wsp1991** [comment]
"they just will have a mandate to not have keys with no
passwords, so rsyslog should support this as an option."
--------->You mean：configure rsyslog  not to encrypted transmission log
, Then there will be no key file.

**Comment by davidelang** [comment]
no, I mean that some companies will say that you can't have a key with no 
password, but that they won't care that you put the password in plaintext in the 
script that uses the key (sometimes in the same directory as the key)

in theory, this protects you in case someone is incompetent enough to steal the 
key but not copy the password along with it.

in practice, if someone is in your system to the point where they can get the 
key, they can also get the password for the key, so encrypting the key doesn't 
really buy you much.

David Lang

**Comment by onet-git** [comment]
The functionality required to address this bug is being able to setup an openssl call back using SSL_CTX_set_default_passwd_cb() so that it can provide the decryption password to the openssl code used to load key files, as described in https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_use_RSAPrivateKey_file.html. 

Unfortunately this is part of a more generic issue of how rsyslog (and other programs) wrap openssl. There are hundreds of different API calls, trying to support them all is infeasible.

**Comment by rgerhards** [comment]
> The functionality required to address this bug

This is not a but - as David said, what does it help if your store the key in just another file? If you can secure that file, you can secure rsyslog configuration (or the key) in the first place. What is the real value of this additional file?

**Comment by rgerhards** [comment]
Sorry for the very late follow-up. This answer may no longer help the original reporter, but it should help operators finding this security question through search. I am not closing this merely because of its age.\n\nAn unattended rsyslog service must obtain the private key without a human prompt. Encrypting the PEM while storing its passphrase in another plaintext config or script only moves the secret and does not improve the host-compromise boundary. The supported deployment model is a root-owned key with restrictive directory/file permissions and least-privilege service access; where a stronger boundary is required, use an external secret/credential mechanism or hardware-backed key support rather than embedding another long-lived passphrase in rsyslog configuration.\n\nBecause this report does not identify a vulnerability in TLS transport and the proposed callback alone would not create a safe unattended secret flow, I am closing this specific request as not planned. A concrete integration proposal for a credential manager or PKCS#11-backed input key would be a distinct enhancement with a meaningful security boundary.

**Comment by rgerhards** [comment]
Deep re-audit correction: the earlier security caveat is valid, but it does not settle the feature request. The original maintainer discussion explicitly recognized password-protected private-key support as useful for compliance deployments, and no implementation or later maintainer rejection was established. Reopening as a credential-loading enhancement, without treating an unencrypted mode-0600 key as a TLS vulnerability.
